Source/WebCore/Modules/gamepad/GamepadHapticActuator.cpp

```diff
@@ -28,34 +28,75 @@
 #if ENABLE(GAMEPAD)
 #include "GamepadHapticActuator.h"
 
+#include "Document.h"
+#include "EventLoop.h"
+#include "Gamepad.h"
 #include "GamepadEffectParameters.h"
+#include "GamepadProvider.h"
 #include "JSDOMPromiseDeferred.h"
+#include <wtf/CompletionHandler.h>
 
 namespace WebCore {
 
-Ref<GamepadHapticActuator> GamepadHapticActuator::create()
+Ref<GamepadHapticActuator> GamepadHapticActuator::create(Gamepad& gamepad)
 {
-    return adoptRef(*new GamepadHapticActuator);
+    return adoptRef(*new GamepadHapticActuator(gamepad));
 }
 
-GamepadHapticActuator::GamepadHapticActuator()
+GamepadHapticActuator::GamepadHapticActuator(Gamepad& gamepad)
     : m_type { Type::Vibration }
+    , m_gamepad { gamepad }
 {
 }
 
-bool GamepadHapticActuator::canPlayEffectType(EffectType) const
+GamepadHapticActuator::~GamepadHapticActuator() = default;
```

Comment:
Shouldn't this be in the header? Or does that create compile issues?

Reply:
This is to avoid unnecessary includes in the header:
```
/Volumes/Work/WebKit/OpenSource/WebKitBuild/Release/usr/local/include/wtf/RefPtr.h:42:16: error: member access into incomplete type 'WebCore::DeferredPromise'
            ptr->deref();
```